### PR TITLE
add remove_var from c++ end

### PR DESCRIPTION
**Pull request by luotao1**
- Similar to `remove_op` method, this PR add `remove_var(var_name)` method for removing a particular varible.
- refine `test_protobuf_desc.py` with https://github.com/PaddlePaddle/Paddle/pull/9600#discussion_r178780837
- refine `block_desc.h` and `protobuf.cc` when `git commit` and `cpplint` fails:
```
paddle/fluid/pybind/protobuf.cc:101:  Is this a non-const reference? If so, make const or use a pointer: T &self  

[runtime/references] [2]
paddle/fluid/pybind/protobuf.cc:110:  Is this a non-const reference? If so, make const or use a pointer: py::module &m  

[runtime/references] [2]
paddle/fluid/pybind/protobuf.cc:154:  Is this a non-const reference? If so, make const or use a pointer: py::module &m  

[runtime/references] [2]
paddle/fluid/pybind/protobuf.cc:215:  Is this a non-const reference? If so, make const or use a pointer: py::module &m  

[runtime/references] [2]
paddle/fluid/pybind/protobuf.cc:266:  Is this a non-const reference? If so, make const or use a pointer: py::module &m  

[runtime/references] [2]
paddle/fluid/pybind/protobuf.cc:130:  Add #include <tuple> for tuple<>  [build/include_what_you_use] [4]
paddle/fluid/pybind/protobuf.cc:304:  Add #include <string> for string  [build/include_what_you_use] [4]
Done processing paddle/fluid/pybind/protobuf.cc
```

### CODE REVIEW

**Comment by wangkuiyi** [comment]
I don't understand why do we need remove_op and remove_var? In my mind, the ProgramDesc is constructive -- built by adding ops and vars from nothing.

**Comment by luotao1** [comment]
@wangkuiyi Sorry for the description of background. The motivation is that we want to fuse batch normalization during inference, and we need an inference transpiler to implement it. As we need to remove batch_norm ops and unused variables from ProgramDesc, we need `remove_op` and `remove_var` methods. Please refer to #9629 for details.

File: paddle/fluid/framework/block_desc.h [hunk 1]
```diff
@@ -96,6 +97,8 @@ class BlockDesc {
    */
   void RemoveOp(size_t s, size_t e);
 
+  void RemoveVar(const std::string &name) { vars_.erase(name); }
```
Comment (typhoonzero) [inline paddle/fluid/framework/block_desc.h]: Note for myself: `sync_with_cpp` already add sync removed vars in previous PR.

File: paddle/fluid/pybind/protobuf.cc [hunk 2]
```diff
@@ -200,13 +202,19 @@ void BindBlockDesc(py::module &m) {
              return self.FindVarRecursive(name);
            },
            py::return_value_policy::reference)
+      .def("remove_var",
```
Comment (typhoonzero) [inline paddle/fluid/pybind/protobuf.cc]: I began to think our previous design of `XXXDesc` maybe a wrong decision. When we need to manipulate the ProgramDesc proto message, we have to write corresponding c++ implementation, then expose it to pybind, and sync the python holders of the c++ side objects, this makes the code complicated and not simple enough.

When comes with transpilers which need to add, cut, remove operators, variables, and blocks in the ProgramDesc, we must be very careful that the c++ side object reference may have already deleted.

A simpler way is to manipulate the ProgramDesc proto message directly in python, and give it to the c++ executor. For performance, python is slow indeed, but if we only use it in compile time and transpile time, it won't affect the total training performance.

Comment (typhoonzero) [reply paddle/fluid/pybind/protobuf.cc]: @wangkuiyi @reyoung can you please take a look at the above comment?

Comment (luotao1) [reply paddle/fluid/pybind/protobuf.cc]: I agree with @typhoonzero. It is so difficult for add, cut, insert, remove operators, variables, and blocks in the ProgramDesc for writing a transpiler, I must change `sync_with_cpp` method in `framework.py` at the same time. 

Related PR: refine sync_with_cpp when remove ops or remove vars https://github.com/PaddlePaddle/Paddle/pull/9600

Comment (panyx0718) [reply paddle/fluid/pybind/protobuf.cc]: Indeed need some rethink
